api/v1_projects.inc

```diff
@@ -271,6 +272,94 @@ function render_project_page_json($row)
     ];
 }
 
+// POST v1/projects/:projectid/rounds/:roundid
+// checkout a new page to proofread
+function api_v1_project_round($method, $data, $query_params)
+{
+    global $pguser;
+
+    $project = $data[":projectid"];
+    $round = $data[":roundid"];
+    check_project_user($project, $round->id);
+
+    try {
+        $lpage = get_available_page($project->projectid, $project->state, $pguser, $err);
+    } catch (Exception $exception) {
+        throw new BadRequest($exception->getMessage());
+    }
+
+    if (is_null($lpage)) {
+        throw new BadRequest($err);
+    }
+    return ["pagename" => $lpage->imagefile, "pagestate" => $lpage->page_state];
+}
+
+// PUT projects/:projectid/pages/:pagename/rounds/:roundid
+// the PUTs all can change the page text or page state.
+// they are idempotent in the sense that if the identical request is issued
+// again, either the the state and text will not be changed or the request will
+// be rejected because the page state has changed so there will be no change
+// to the server. Logging side-effects excluded.
+function api_v1_project_page_round($method, $data, $query_params)
+{
+    global $pguser;
+
+    $project = $data[":projectid"];
+    $round = $data[":roundid"];
+    check_project_user($project, $round->id);
+
+    try {
+        $lpage = get_indicated_LPage($project->projectid, $project->state, $data[":pagename"], $query_params["pagestate"], 0);
+
+        // can throw InvalidArgumentException
+        $action = get_enumerated_param($query_params, "action", null, ["return", "save_temp", "save_done", "open", "revert"]);
+
+        switch ($action) {
+        case "return":
+            $lpage->returnToRound($pguser);
+            return [];
+            break;
+        case "save_temp":
+            $lpage->saveAsInProgress(api_get_request_body(), $pguser);
+            return render_page_data($lpage);
+            break;
+        case "save_done":
+            [$page_saved, $limit_reached] = $lpage->attemptSaveAsDone(api_get_request_body(), $pguser);
+            return ["pageSaved" => $page_saved, "limitReached" => $limit_reached];
+        case "open":
+            $lpage->resume_saved_page($pguser);
+            return render_page_data($lpage);
+            break;
+        case "revert":
```

Comment:
Reverting is an interface-level artifact that shouldn't be in LPage to begin with (LPage even says as much on line 240). If the user wants to revert, the client should save the page as in-progress, GET the page text from the prior round, and present that to the user.

Reply:
The way it is now fits quite well onto how LPage works. Should we change the way LPage works first if we want to do this?

Reply:
Which is to say leave it as-is but don't use it in the API. We need to design and use the API in ways that make sense now while enabling the functionality we need in a modern interface.